### PR TITLE
--bless all mir-opt tests.

### PR DESCRIPTION
**Pull request by anyska**
r? @oli-obk

### CODE REVIEW

File: src/test/mir-opt/byte_slice/rustc.main.SimplifyCfg-elaborate-drops.after.mir [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        _0 = ();                         // bb0[4]: scope 0 at $DIR/byte_slice.rs:4:11: 7:2
+        StorageDead(_2);                 // bb0[5]: scope 1 at $DIR/byte_slice.rs:7:1: 7:2
+        StorageDead(_1);                 // bb0[6]: scope 0 at $DIR/byte_slice.rs:7:1: 7:2
+        return;                          // bb0[7]: scope 0 at $DIR/byte_slice.rs:7:2: 7:2
+    }
+}
+
+alloc0 (size: 3, align: 1) {
+    66 6f 6f                                        │ foo
+}
```
Comment (eddyb) [inline src/test/mir-opt/byte_slice/rustc.main.SimplifyCfg-elaborate-drops.after.mir:46]: Interesting outcome: the `ty::Const` uses this, but it also gets pretty-printed, so the `alloc0` isn't strictly necessary.
But at least the super-verbose comments do mention `Scalar(alloc0+0)`, so at least for now it doesn't hurt.

File: src/test/mir-opt/graphviz/rustc.main.mir_map.0.dot [hunk 2]
```diff
@@ -0,0 +1,10 @@
+digraph Mir_0_3 {
+    graph [fontname="monospace"];
+    node [fontname="monospace"];
+    edge [fontname="monospace"];
+    label=<fn main() -&gt; ()<br align="left"/>>;
+    bb0__0_3 [shape="none", label=<<table border="0" cellborder="1" cellspacing="0"><tr><td bgcolor="gray" align="center" colspan="1">0</td></tr><tr><td align="left" balign="left">_0 = ()<br/></td></tr><tr><td align="left">goto</td></tr></table>>];
+    bb1__0_3 [shape="none", label=<<table border="0" cellborder="1" cellspacing="0"><tr><td bgcolor="gray" align="center" colspan="1">1</td></tr><tr><td align="left">resume</td></tr></table>>];
+    bb2__0_3 [shape="none", label=<<table border="0" cellborder="1" cellspacing="0"><tr><td bgcolor="gray" align="center" colspan="1">2</td></tr><tr><td align="left">return</td></tr></table>>];
+    bb0__0_3 -> bb2__0_3 [label=""];
+}
```
Comment (eddyb) [inline src/test/mir-opt/graphviz/rustc.main.mir_map.0.dot:10]: Whoa! I didn't even know `mir-opt` tests could test for this.
I now wish GitHub could render this as a graph :disappointed:.

File: src/test/mir-opt/address-of/rustc.address_of_reborrow.SimplifyCfg-initial.after.mir [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                                         // + ty: i32
+                                         // + val: Value(Scalar(0x00000000))
+                                         // mir::Constant
+                                         // + span: $DIR/address-of.rs:5:23: 5:24
+                                         // + literal: Const { ty: i32, val: Value(Scalar(0x00000000)) }
+        _3 = &mut _4;                    // bb0[8]: scope 1 at $DIR/address-of.rs:5:17: 5:29
+        FakeRead(ForLet, _3);            // bb0[9]: scope 1 at $DIR/address-of.rs:5:9: 5:14
+        StorageLive(_5);                 // bb0[10]: scope 2 at $DIR/address-of.rs:7:5: 7:18
+        StorageLive(_6);                 // bb0[11]: scope 2 at $DIR/address-of.rs:7:5: 7:18
+        _6 = &raw const (*_1);           // bb0[12]: scope 2 at $DIR/address-of.rs:7:5: 7:6
```
Comment (oli-obk) [inline src/test/mir-opt/address-of/rustc.address_of_reborrow.SimplifyCfg-initial.after.mir:153]: This was `_5 = &raw const (*_1)` in the mir opt test, why did it change when moving to --bless tests?

File: src/test/mir-opt/address-of.rs [hunk 4]
```diff
@@ -35,78 +37,11 @@ fn address_of_reborrow() {
 }
 
 // The normal borrows here should be preserved
+// EMIT_MIR rustc.borrow_and_cast.SimplifyCfg-initial.after.mir
```
Comment (oli-obk) [inline src/test/mir-opt/address-of.rs:40]: The previous tests were checking for `rustc.borrow_and_cast.EraseRegions.after.mir`

File: src/test/mir-opt/retag.rs [hunk 5]
```diff
@@ -16,10 +18,14 @@ impl Test {
     }
 }
 
+//rustc.ptr-drop_in_place.Test.SimplifyCfg-make_shim.after.mir
```
Comment (oli-obk) [inline src/test/mir-opt/retag.rs:21]: This comment doesn't emit anything

File: src/test/mir-opt/simplify_if.rs [hunk 6]
```diff
@@ -1,19 +1,7 @@
+// EMIT_MIR rustc.main.SimplifyBranches-after-const-prop.diff
+
 fn main() {
     if false {
         println!("hello world!");
```
Comment (oli-obk) [inline src/test/mir-opt/simplify_if.rs:5]: we should probably make this test less noisy by doing something trivial here like `std::process::exit(1);`

File: src/test/mir-opt/simplify_match.rs [hunk 7]
```diff
@@ -1,22 +1,8 @@
+// EMIT_MIR rustc.main.SimplifyBranches-after-copy-prop.diff
+
 fn main() {
     match { let x = false; x } {
         true => println!("hello world!"),
```
Comment (oli-obk) [inline src/test/mir-opt/simplify_match.rs:5]: same here wrt noisyness

File: src/test/mir-opt/uninhabited_enum_branching.rs [hunk 8]
```diff
@@ -14,6 +14,8 @@ enum Test2 {
     E = 5,
 }
 
+// EMIT_MIR rustc.main.UninhabitedEnumBranching.diff
+// EMIT_MIR rustc.main.SimplifyCfg-after-uninhabited-enum-branching.after.mir
```
Comment (oli-obk) [inline src/test/mir-opt/uninhabited_enum_branching.rs:18]: I've seen this pattern a few times. I wonder if we should have a diff mode for "before Optimization X to after optimization Y"

File: src/test/mir-opt/unusual-item-types.rs [hunk 9]
```diff
@@ -1,86 +1,27 @@
...
 }
 
+// EMIT_MIR rustc.E-V-{{constant}}.mir_map.0.mir
 enum E {
     V = 5,
 }
 
 fn main() {
     let f = Test::X as fn(usize) -> Test;
+    // EMIT_MIR rustc.ptr-drop_in_place.std__vec__Vec_i32_.AddMovesForPackedDrops.before.mir
```
Comment (oli-obk) [inline src/test/mir-opt/unusual-item-types.rs:25]: there was no mir dump emitted for this comment. I think `--bless` only detects comments at the start of the line... not indented ones. I should fix that XD

Comment (oli-obk) [reply src/test/mir-opt/unusual-item-types.rs:25]: Please just unindent it for now

File: src/test/mir-opt/address-of.rs (hunk 4 above)
Comment (oli-obk) [reply src/test/mir-opt/address-of.rs:40]: nevermind, `EraseRegions` is gone

File: src/test/mir-opt/unusual-item-types.rs (hunk 9 above)
Comment (oli-obk) [reply src/test/mir-opt/unusual-item-types.rs:25]: Please also check whether I missed other instances of this

**Comment by eddyb** [comment]
So that last failure looks like a `compiletest` normalization bug: only the "start" `line:column` are changed to `LL:COL`, the "end" ones aren't:
```diff
-                   debug t => _6;           // in scope 7 at $SRC_DIR/libcore/convert/mod.rs:LL:COL: 566:14
+                   debug t => _6;           // in scope 7 at $SRC_DIR/libcore/convert/mod.rs:LL:COL: 565:14
```

**Comment by oli-obk** [comment]
why is there any normalization? We don't normalize mir-opt dump files

**Comment by eddyb** [comment]
@anyska Btw, once there are no files remaining with `START`/`END` markers, you can remove:
https://github.com/rust-lang/rust/blob/1b521f57735663de9373679cf8c6502622036bf1/src/tools/compiletest/src/runtest.rs#L3160-L3188

@oli-obk Looking at compiletest, something we all forgot about is that right now there's no error if there's no `// EMIT_MIR` or `// END RUST SOURCE` in the entire file. I believe they may have been untested tests which were missing `// END RUST SOURCE`.

**Comment by oli-obk** [comment]
@bors r+ p=1

let's get this landed quickly and address all the mir opt improvements later

**Comment by eddyb** [comment]
cc @RalfJung [if you look at the log](https://dev.azure.com/rust-lang/rust/_build/results?buildId=25362&view=logs&j=604eab81-0cf0-52c5-10f8-91bf5e7241d2&t=8044b509-45d7-5e81-e94a-52c792e952fd&l=16553), all of the failures are because `Scalar`'s debug printout pads `usize` so the full width.
So we'll have to use the "for each bit width" feature even if the difference is more zeros on a single line.

**Comment by RalfJung** [comment]
> So we'll have to use the "for each bit width" feature even if the difference is more zeros on a single line.

I don't know what that feature is, but maybe there is a way to normalize this away?
Like, it could replace `0x0000000000` by `0x00`, that should have few false positives and cover most things.

**Comment by eddyb** [comment]
This is happening again: https://github.com/rust-lang/rust/pull/70721#issuecomment-608376443.

**Comment by oli-obk** [comment]
I'm on it

File: src/test/mir-opt/const_prop/issue-67019/rustc.main.ConstProp.diff [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                                           // mir::Constant
+-                                          // + span: $DIR/issue-67019.rs:11:12: 11:13
++                                          // + span: $DIR/issue-67019.rs:11:11: 11:17
+                                           // + literal: Const { ty: u8, val: Value(Scalar(0x01)) }
+                                           // ty::Const
+                                           // + ty: u8
+                                           // + val: Value(Scalar(0x02))
+                                           // mir::Constant
+-                                          // + span: $DIR/issue-67019.rs:11:15: 11:16
++                                          // + span: $DIR/issue-67019.rs:11:11: 11:17
```
Comment (eddyb) [inline src/test/mir-opt/const_prop/issue-67019/rustc.main.ConstProp.diff:27]: @wesleywiser @oli-obk What is happening here, why is the span of the `mir::Constant` changing?

Comment (wesleywiser) [reply src/test/mir-opt/const_prop/issue-67019/rustc.main.ConstProp.diff:27]: I suspect [`ConstPropagator::replace_with_const()`](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/const_prop.rs#L638) is getting called and is using the `Span` from the MIR `Statement` which creates a new `Const` and replaces the existing one with the correct `Span`. We probably just need to early return out of `replace_with_const()` if `rvalue` is already a `Const`.

File: src/test/mir-opt/inline/inline-closure-borrows-arg.rs [hunk 11]
```diff
@@ -7,53 +7,11 @@ fn main() {
     println!("{}", foo(0, &14));
 }
 
+// EMIT_MIR rustc.foo.Inline.after.mir
```
Comment (eddyb) [inline src/test/mir-opt/inline/inline-closure-borrows-arg.rs:10]: Okay looks like inline tests all have this problem, so I'm leaning in favor of a separate PR.

**Review by eddyb — CHANGES_REQUESTED** [review]
(not done yet, trying to reset the GitHub UI to something sensible)

File: src/test/mir-opt/nll/region-subtyping-basic.rs [hunk 12]
```diff
@@ -19,23 +21,3 @@ fn main() {
...
-// debug p => _2;
-// ...
-// let _6: &'_#4r usize;
-// ...
-// debug q => _6;
-// ...
-// _2 = &'_#2r _1[_3];
-// ...
-// _6 = _2;
-// END rustc.main.nll.0.mir
```
Comment (eddyb) [inline src/test/mir-opt/nll/region-subtyping-basic.rs:41]: cc @nikomatsakis @matthewjasper are you okay with changing NLL tests to the new system that contains the entire MIR output?

Comment (nikomatsakis) [reply src/test/mir-opt/nll/region-subtyping-basic.rs:41]: I'm ok with that. Honestly, we could delete those particular tests, they were useful in getting started but I'm not sure they provide much value now.

Comment (eddyb) [reply src/test/mir-opt/nll/region-subtyping-basic.rs:41]: I still think they're useful as canaries, to be able to detect when something changes.
Maybe useful for polonius?

File: src/test/mir-opt/uninhabited_enum_branching.rs (hunk 8 above)
Comment (eddyb) [reply src/test/mir-opt/uninhabited_enum_branching.rs:18]: Maybe we should always use `.diff`.

**Review by eddyb — APPROVED** [review]

File: src/test/mir-opt/uninhabited_enum_branching.rs (hunk 8 above)
Comment (oli-obk) [reply src/test/mir-opt/uninhabited_enum_branching.rs:18]: What I mean is to diff across multiple optimizations instead of just a single one.

**Comment by eddyb** [comment]
[Failures](https://dev.azure.com/rust-lang/rust/_build/results?buildId=25753&view=logs&j=7f477069-cc2e-5bb5-c142-f030a79bc52a&t=98b75561-f476-54d2-b1f7-94cdb0af816f&l=16754):
```
    [mir-opt] mir-opt/issue-41888.rs
    [mir-opt] mir-opt/match-arm-scopes.rs
```

I can't easily tell if non-determinism/sensitivity to test conditions is the problem, or a rebase is needed.

**Comment by Dylan-DPC** [comment]
deprioritising @bors p=1

**Pull request merged**